File: src/fsharp/Agent.fs
module Actors.Agents

open System

type Agent<'T> = MailboxProcessor<'T>

type message = End | Rectangle of int *int | Circle of float

let start() =
    Agent.Start(fun inbox ->
    let rec loop () = async {
        let! msg = inbox.Receive()
        match msg with
        | End ->
            printf "Bye \n"
            return()
        | Rectangle(width, height) ->
            printf "Area is %O\n" (width * height)
            return! loop()
        | Circle(radius) ->
            printf "Area is %O \n" (Math.PI * radius * radius)
            return! loop()
    }
    loop())

File: src/scala/Agent.scala
package scala

import actors.Actor

class Agent extends Actor {
  def act() {
    loop {
      receive {
        case ('hello, x: String) =>
          println("Hello " + x: String)
        case ('end) =>
          exit('ok)
        case _ =>
          println("WTF")
      }
    }
  }
}

object Main extends App {
  val agent = new Agent().start()
  agent !('hello, "My message")
  agent ! 42
  agent ! ('end)
}
